core/src/main/java/com/scalar/db/common/TransactionDecorationTwoPhaseCommitTransactionManager.java

```diff
@@ -0,0 +1,224 @@
+package com.scalar.db.common;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.scalar.db.api.Delete;
+import com.scalar.db.api.Get;
+import com.scalar.db.api.Insert;
+import com.scalar.db.api.Mutation;
+import com.scalar.db.api.Put;
+import com.scalar.db.api.Result;
+import com.scalar.db.api.Scan;
+import com.scalar.db.api.TwoPhaseCommitTransaction;
+import com.scalar.db.api.Update;
+import com.scalar.db.api.Upsert;
+import com.scalar.db.common.error.CoreError;
+import com.scalar.db.config.DatabaseConfig;
+import com.scalar.db.exception.transaction.AbortException;
+import com.scalar.db.exception.transaction.CommitException;
+import com.scalar.db.exception.transaction.CrudException;
+import com.scalar.db.exception.transaction.PreparationException;
+import com.scalar.db.exception.transaction.RollbackException;
+import com.scalar.db.exception.transaction.TransactionException;
+import com.scalar.db.exception.transaction.UnknownTransactionStatusException;
+import com.scalar.db.exception.transaction.ValidationException;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public abstract class TransactionDecorationTwoPhaseCommitTransactionManager
+    extends AbstractTwoPhaseCommitTransactionManager
+    implements TwoPhaseCommitTransactionDecoratorAddable {
+
+  private final List<TwoPhaseCommitTransactionDecorator> transactionDecorators =
+      new CopyOnWriteArrayList<>();
+
+  public TransactionDecorationTwoPhaseCommitTransactionManager(DatabaseConfig config) {
+    super(config);
+
+    // Add the StateManagedTransaction decorator by default
+    addTransactionDecorator(StateManagedTransaction::new);
+  }
+
+  protected TwoPhaseCommitTransaction decorate(TwoPhaseCommitTransaction transaction)
+      throws TransactionException {
+    TwoPhaseCommitTransaction decorated = transaction;
+    for (TwoPhaseCommitTransactionDecorator transactionDecorator : transactionDecorators) {
+      decorated = transactionDecorator.decorate(decorated);
+    }
+    return decorated;
+  }
+
+  @Override
+  public void addTransactionDecorator(TwoPhaseCommitTransactionDecorator transactionDecorator) {
+    transactionDecorators.add(transactionDecorator);
+  }
+
+  /**
+   * This class is to unify the call sequence of the transaction object. It doesn't care about the
+   * potential inconsistency between the status field on JVM memory and the underlying persistent
+   * layer.
+   */
+  @VisibleForTesting
+  static class StateManagedTransaction extends DecoratedTwoPhaseCommitTransaction {
+
+    private enum Status {
+      ACTIVE,
```

Comment:
[minor] This is a bit off topic. But `PREPARED` and `VALIDATED` may seem active as well. So, `EXECUTING` or `CRUD` or something might be less confusing?

Reply:
Yes, you are right. Let's discuss this in a separate PR. Thanks.